src/aws/flb_aws_util.c

```diff
@@ -507,3 +512,182 @@ int flb_read_file(const char *path, char **out_buf, size_t *out_size)
 
     return 0;
 }
+
+/* Generic replace function for strings. */
+static char* replace_uri_tokens(const char* original_string, const char* current_word, 
+                         const char* new_word) 
+{ 
+    char *result; 
+    int i = 0;
+    int count = 0; 
+    int new_word_len = strlen(new_word); 
+    int old_word_len = strlen(current_word); 
+  
+    for (i = 0; original_string[i] != '\0'; i++) { 
+        if (strstr(&original_string[i], current_word) == &original_string[i]) { 
+            count++; 
+            i += old_word_len - 1; 
+        } 
+    }
+
+    result = flb_sds_create_size(i + count * (new_word_len - old_word_len) + 1);
+    if (!result) {
+        flb_errno();
+        return NULL;
+    }
+
+    i = 0; 
+    while (*original_string) { 
+        if (strstr(original_string, current_word) == original_string) { 
+            strcpy(&result[i], new_word); 
+            i += new_word_len; 
+            original_string += old_word_len; 
+        } 
+        else
+            result[i++] = *original_string++; 
+    } 
+
+    result[i] = '\0'; 
+    return result; 
+} 
+
+/* Constructs S3 object key as per the format. */
+flb_sds_t flb_get_s3_key(const char *format, time_t time, char *tag, char *tag_delimiter)
+{
+    int i = 0;
+    int ret = 0;
+    char *tag_token = NULL;
+    flb_sds_t tmp = NULL;
+    flb_sds_t buf = NULL;
+    flb_sds_t s3_key = NULL;
+    flb_sds_t tmp_key = NULL;
+    flb_sds_t tmp_tag = NULL;
+    struct tm *gmt = NULL;
+    
+    if (strlen(format) > S3_KEY_SIZE){
+        flb_warn("[s3_key] Object key length is longer than the 1024 character limit.");
+    }
+
+    tmp_tag = flb_sds_create_len(tag, strlen(tag));
+    if(!tmp_tag){
+        goto error;
+    }
+
+    s3_key = flb_sds_create_len(format, strlen(format));
+    if (!s3_key) {
+        goto error;
+    }
+
+    /* Check if delimiter(s) specifed exists in the tag. */
+    for (i = 0; i < strlen(tag_delimiter); i++){
+        if (strchr(tag, tag_delimiter[i])){
+            ret = 1;
+            break;
+        } 
+    }
+
+    if(ret == 0){
+        flb_warn("[s3_key] Invalid Tag delimiter: does not exist in tag. "
+                 "tag=%s, format=%s", tag, format);
+    }  
```

Comment:
Was testing, and this warning message is printed in cases that we don't want it to:

```
[2020/09/09 06:43:50] [ warn] [s3_key] Invalid Tag delimiter: does not exist in tag. tag=139ceeecd368, format=fluent-bit-logs/$TAG/%Y/%m/%d/%H/%M/%S
```

In that case, the tag did not contain the delimiter, but the format did not contain any `$TAG[i]`, so the warning is not useful and might confuse the user.

Add a check for tag parts used in format around this warning:

```
if (strstr(format, "$TAG[") != NULL) {
...
}
```